state/store.go

```diff
@@ -212,133 +229,168 @@ func (store dbStore) Bootstrap(state State) error {
 	return store.db.SetSync(stateKey, state.Bytes())
 }
 
-// PruneStates deletes states between the given heights (including from, excluding to). It is not
+// PruneStates deletes states up to the height specified (exclusive). It is not
 // guaranteed to delete all states, since the last checkpointed state and states being pointed to by
-// e.g. `LastHeightChanged` must remain. The state at to must also exist.
-//
-// The from parameter is necessary since we can't do a key scan in a performant way due to the key
-// encoding not preserving ordering: https://github.com/tendermint/tendermint/issues/4567
-// This will cause some old states to be left behind when doing incremental partial prunes,
-// specifically older checkpoints and LastHeightChanged targets.
-func (store dbStore) PruneStates(from int64, to int64) error {
-	if from <= 0 || to <= 0 {
-		return fmt.Errorf("from height %v and to height %v must be greater than 0", from, to)
-	}
-	if from >= to {
-		return fmt.Errorf("from height %v must be lower than to height %v", from, to)
-	}
-	valInfo, err := loadValidatorsInfo(store.db, to)
-	if err != nil {
-		return fmt.Errorf("validators at height %v not found: %w", to, err)
+// e.g. `LastHeightChanged` must remain. The state at retain height must also exist.
+// Pruning is done in descending order.
+func (store dbStore) PruneStates(retainHeight int64) error {
+	if retainHeight <= 0 {
+		return fmt.Errorf("height %v must be greater than 0", retainHeight)
+	}
+
+	if err := store.pruneValidatorSets(retainHeight); err != nil {
+		return err
+	}
+
+	if err := store.pruneConsensusParams(retainHeight); err != nil {
+		return err
 	}
-	paramsInfo, err := store.loadConsensusParamsInfo(to)
+
+	if err := store.pruneABCIResponses(retainHeight); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (store dbStore) pruneValidatorSets(height int64) error {
+	valInfo, err := loadValidatorsInfo(store.db, height)
 	if err != nil {
-		return fmt.Errorf("consensus params at height %v not found: %w", to, err)
+		return fmt.Errorf("validators at height %v not found: %w", height, err)
+	}
+
+	// We will prune up to the validator set at the given "height". As we don't save validator sets every
+	// height but only when they change or at a check point, it is likely that the validator set at the height
+	// we prune to is empty and thus dependent on the validator set saved at a previous height. We must find
+	// that validator set and make sure it is not pruned by saving after we have finished pruning.
```

Comment:
I don't think this comment is valid anymore.

Reply:
I think most of it is still valid. We still keep, in most cases, one validator set that the validator set at retain height is dependent on. The last bit is now outdated as we just skip over this one

Reply:
Ah yeah, I was referring to the last bit, sorry.